README.md
# FluidPage:

## A Brief Introduction to Creating a TYPO3 site with FluidPage


### About FluidPage

This extension was developed by the good people at Cast Iron Coding. It was created out of the impression that the two common solutions to templating in TYPO3, Templavoila and AutoMakeTemplate, were both insufficient. Templavoila tends to not store data in a normalized format because of its heavy reliance on XML. Indeed, it tends to obfuscate the relationship between pages and page content, which can cause problems on many kinds of TYPO3 projects. Moreover, the UI of Templavoila has often been somewhat out of sync with UI of TYPO3 in general, which we believe causes problems for users. The Typoscript-driven approach to templating, which is simplified somewhat by AutoMakeTemplate, is time consuming. Whenever a new element is added to the page, TypoScript has to be written so that objects are mapped to the correct spot in the template. We find this approach to be rather tedious. Moreover, the long-standing problem with classic TYPO3 templating has been the limitation of the 4 columns in the page module: left, normal, right and border.

Recent versions of TYPO3 ship with Fluid and backend-layouts. Fluid, as you no doubt know if you're reading this, is a powerful templating engine. There is no reason why it should only be usable in TYPO3 extensions, and we believe it is time to fully embrace it in the TYPO3 core and use it as _the_ core templating mechanism. Backend layouts solve the important problem of the UI in the page module by allowing admins the flexibility to create complex layouts and a corresponding editing UI. In the near future, we hope to see the gridelements extension in the TYPO3 core, which will address the need for nested content elements (currently only really met by TemplaVoila) in the page module.

This extension is still rather alpha. We are using it in multiple production sites and are improving the extension on a daily basis. This extension is, thankfully, a simple extension. Instead of trying to provide complex functionality, it tries to provide a sensible way to tie together existing aspects of TYPO3 (backend layouts, fluid templates, and typoscript). If you take a look at the pi1 script that ships with this extension, you will see that it is quite simple, and we hope it remains that way. With simplicity comes stability. Simple as it may be, we believe this extension opens up a lot of doors in TYPO3 and brings real power to the TYPO3 site implementation process.

### FluidPage Development

We are currently updating this extension on Github. I'll be honest. Here at CIC, Git is so wound up in our process it's difficult for us to think about hosting this in SVN on forge.typo3.org. If people start using this extension and demand picks up, we may re-evaluate that decision. Once we feel the extension is stable, we will release a version to the TER. We're highly committed to this approach to building sites with TYPO3; if you find an issue, we will do our very best to fix it promptly. Please report issues on the github page at [https://github.com/castiron/fluidpage/issues](https://github.com/castiron/fluidpage/issues).

##### 1. Install the fluidpage extension

The simplest way to do this, for now, is to clone the repository using git, although you can also install the extension from the TER. To clone it from github, execute this command in the typo3conf/ext directory:

```
git clone git@github.com:castiron/fluidpage.git fluidpage
```

Once that's done, you can install the extension in the extension manager as with any extension. The FluidPage extension does not add any database fields to TYPO3\. However, it does make some minor changes to the TCA by way of TSconfig. Out of the box, fluid page will change the label on the "backend layout" field on page records to "Page Template (this page only)". The child page backend layout field will be changed to "Page Template (subpages of this page)". It makes this UI change in order to make it more clear to content editors, who may not understand the concept of a backend layout, that these layout fields are used to select the template that will be used in both the backend and the frontend.

##### 2. Setup the page tree

To work through this tutorial, you'll need a simple page tree as a starting point. Create a new root page and beneath it three or four additional pages. Create a folder to store your backend layouts.

##### 3. Create a back-end layout

In the storage folder you created in the previous step, make a back-end layout record. For the purposes of this how-to, go ahead and create a backend layout record with two columns called left and right. Assign a column number of "1" to the left column and a column number of "2" to the right column. Your layout configuration should look more or less like this when you're done:

```
backend_layout {
        colCount = 2
        rowCount = 1
        rows {
                1 {
                        columns {
                                1 {
                                        name = Left Column
                                        colPos = 1
                                }
                                2 {
                                        name = Right Column
                                        colPos = 2
                                }
                        }
                }
        }
}
```

Set the title of the backend layout to "two column" and save the record.

##### 4. Assign your back-end layout to pages in the page tree

Edit the root page in the page tree. You'll see two fields on the "appearance" tab for selecting the page template. Set both of them to your newly created back-end layout.

##### 5. Create a Typoscript template

On your root page (or wherever you like to put your master Typoscript template) create a new template. Edit the template and switch to the "include" tab. Include CSS Styled Content (of course) and the "Fluid Page Static Typoscript Template" options. Edit the setup field of the template and add the following:

```
lib.primaryNav = HMENU
lib.primaryNav {
        entryLevel = 0
        wrap = <ul id="topNav" class="nav">|</ul>
        1 = TMENU
        1.expAll = 1
        1.noBlur = 1
        1.NO {
                stdWrap.htmlSpecialChars = 1
                wrapItemAndSub = <li class="first">|</li>|*|<li>|</li>|*|<li class="last">|</li>
        }
        1.ACT = 1
        1.ACT < .1.NO
        1.ACT {
                ATagParams = class="active"
        }
}

lib.secondaryNav < lib.primaryNav
lib.secondaryNav {
        wrap >
        1.expAll = 0
        1.wrap = <ul class="nav">|</ul>
        2.expAll = 0
}

lib.content.leftCol < styles.content.get
lib.content.leftCol.select.where = colPos=1

lib.content.rightCol < styles.content.get
lib.content.rightCol.select.where = colPos=2
lib.content.rightCol.slide = -1

page = PAGE
page {
        10 < plugin.tx_fluidpage_pi1
        10 {
                settings {
                        partialRootPath = EXT:fluidpage/Resources/Public/Partials/
                layoutRootPath = EXT:fluidpage/Resources/Public/Layouts/
                variables {
                        tagline  = TEXT
                        tagline {
                                        value = This tagline on page "{field:title}" is an example of a global template variable
                                        insertData = 1
                                        stdWrap = <strong>|</strong>
                        }
                }
                }
                templates {
                        1 {
                                file = EXT:fluidpage/Resources/Public/html/main.html
                                constants {
                                        showLeftColumn = 1
                                        showRightColumn = 1
                                        bodyClass = column-2
                                }
                                variables {
                                        tagline  = TEXT
                                tagline {
                                                value = Example: template variable overriding a global variables on page "{field:title}."
                                                insertData = 1
                                                stdWrap = <strong>|</strong>
                                }
                                }
                        }
                        2 {
                                file = EXT:fluidpage/Resources/Public/html/main.html
                                constants {
                                        showLeftColumn = 1
                                        showRightColumn = 0
                                        bodyClass = column-1
                                }
                        }
                }
        }

        typeNum = 0

        # Unset the body tag, as it's is provided by the fluid template
        bodyTagCObject = TEXT
        bodyTagCObject.noTrimWrap = | | |

        includeCSS {
                10 = typo3conf/ext/fluidpage/Resources/Public/css/style.css
        }
}
```

It's worth pausing for a moment here to review what's happening in this typoscript. The first two objects are `lib.primaryNav` and `lib.secondaryNav`. These are simple `HMENU` objects that are used to render the primary and secondary navigation. I've included them here so that you can see how you can reference a typoscript object in a fluidPage template using a fluid view helper (eg: `<f:CObject typoscriptObjectPath="lib.secondaryNav" />`). The ability to reference Typoscript CObjects directly in your fluidPage template is an important part of this approach because it means there is no longer a mapping step, as there was in TemplaVoila and in AutoMakeTemplate. Instead of mapping Typoscript objects to templates, you simply reference them directly in your template by way of view helpers.

After the menu objects, there are some lib.content declarations. In this section, we're simple declaring the various content areas that are available in the backend. Remember when you set the left column to have a colPos of 1? This is where we reference that column position and assign the rendered content for that column to a typoscript object, which can in turn be rendered in the fluid template just like the primary or secondary navigation. Notice that the right column is configured to slide down the page tree, using existing Typoscript properties.

In the page declaration we tell TYPO3 to take the output of `plugin.tx_fluidpage_pi1` and assign it to `page.10`. This is similar to how TemplaVoila works. The "templates" section of the fluidPage configuration is the key part. Each item under templates (1, 2, etc) refers to the UID of a back-end layout. Let's look at one template declaration:

```
templates {
        1 {
                file = EXT:fluidpage/Resources/Public/html/main.html
                constants {
                        showLeftColumn = 1
                        showRightColumn = 1
                }
        }
}
```

This tells FluidPage that when the backend layout for a given page is the backend layout record with the UID of 1, it should use the fluid template file at `EXT:fluidpage/Resources/html/main.html`. The path to this file could, of course, be a file in `fileadmin/templates`. At Cast Iron Coding we prefer to store all template assets in an extension, so we've made sure that FluidPage can parse paths beginning with EXT.

The constants section contains arbitrary constants declarations which can be referenced in the fluid template as values in the {constants} array. For example, a fluid template could contain a condition that referenced a constant:

```
<f:if condition="{constants.showLeftColumn} == 1">
<div id="main">
        <f:CObject typoscriptObjectPath="lib.breadcrumb" />
        <div id="mainContent">
                <h2>Left Column</h2>
                <f:CObject typoscriptObjectPath="lib.content.leftCol" />
        </div>
</div>
```

Constants make it possible to have, for example, a single template with sections that show up depending on how a specific template is configured. Constants also make it easier to re-use template functionality in multiple backend layouts.

To see how the actual fluid template works, take a look at the file at `EXT:fluidpage/Resources/Public/html/main.html`. Notice that the page record is available to the fluid view as `{page}`. This makes it possible to include the page title, for example, in the header by simply adding `{page.title}` to the fluid template. For light-weight rendering tasks like this, there's no need to create a whole typoscript object.

ChangeLog
commit 5a789c23ee3ad35effc96d57b395d7ad7232e198
Author: Zach Davis <zach@castironcoding.com>
Date:   Mon Feb 20 12:10:47 2012 -0800

    [CHORE] Removed .DS_Store file and updated .gitignore

commit e3d14b3c5c5270e84b2a95904e1dbb71b02e025c
Merge: d91ec4d 79e7f2f
Author: Zach Davis <zach@castironcoding.com>
Date:   Mon Feb 20 12:06:10 2012 -0800

    Merge branch 'master' of git://github.com/Tryweb/fluidpage

commit d91ec4df21244a44d8f8368e96ac11513f300e56
Author: Zach Davis <zach@castironcoding.com>
Date:   Mon Feb 20 11:55:24 2012 -0800

    [FEATURE] Added suppport for Fluid Layouts back into FluidPages. Layoutpath can now be set in Typoscript and the extension will respect any layout tags that are present within the ###layout### subpart marker

commit 79e7f2f7bdfec4d92bb242c328d5f1db2c7cd069
Author: Sebastiaan van Parijs <svparijs@tryweb.nl>
Date:   Fri Feb 10 14:15:11 2012 +0100

    Removed warning when using News plugin

commit 1539c945f4a9e28da88c34b52917a6a92a85da82
Author: Zach Davis <zach@castironcoding.com>
Date:   Tue Dec 13 11:33:21 2011 -0800

    [BUGFIX] Removed the addPlugin call in ext_tables.php; Fluidpage does not contain a plugin that can be inserted on a page as a content element

commit f3c22e12c7ff5d809172702aa79e63d000afe6ed
Merge: 8fcc92c b80bb90
Author: Zach Davis <zach@castironcoding.com>
Date:   Fri Nov 18 16:00:01 2011 -0800

    Merge branch 'master' of github.com:castiron/fluidpage

commit 8fcc92cf43534744f039973cac0c1d2a6a954039
Author: Zach Davis <zach@castironcoding.com>
Date:   Fri Nov 18 11:42:22 2011 -0800

    [CHORE] Released 0.2.0 to TER

commit b80bb90ad993e0ed76671db97825028f7a67964f
Author: lthurston <lucas@castironcoding.com>
Date:   Tue Nov 8 10:38:57 2011 -0800

    Modified documentation to describe the new Fluidpage content viewhelper.

commit b7381f0e0428d3b8ebe55385a91fed7d64a3079e
Author: Lucas Thurston <lucas@castironcoding.com>
Date:   Fri Nov 4 14:19:57 2011 -0700

    [FEATURE] Content viewhelper created

commit c0f5b06f1dce7f512d57e3cfc7d368dfb2d2eb9d
Merge: f0e559d 697e471
Author: Michael McManus <michaelorionmcmanus@gmail.com>
Date:   Fri Oct 28 12:09:28 2011 -0700

    Merge branch 'master' of github.com:castiron/fluidpage

commit f0e559d1d4d0bfaeb5d3236ee0696681a6ae6b87
Author: Michael McManus <michaelorionmcmanus@gmail.com>
Date:   Fri Oct 28 12:09:11 2011 -0700

    [CHORE] Modified permissions on files within the Classes directory.

commit 697e471faa89ad1c890ecb5b62041c723b8fbbe4
Author: Zach Davis <zach@castironcoding.com>
Date:   Fri Oct 28 07:35:52 2011 -0700

    [BUGFIX] Corrected path to the Classes directory in ext_autoload

commit c6974c1c3a6eed69f308e126252ece9fe1a1a7b1
Author: Michael McManus <michaelorionmcmanus@gmail.com>
Date:   Thu Oct 27 16:37:46 2011 -0700

    [FEATURE] Process of adding backend_layout page fields to the front end rootline is more robust.

commit 8c74517f0607afde31586c799a8684f78347270b
Author: Zach Davis <zach@castironcoding.com>
Date:   Fri Oct 21 12:03:01 2011 -0700

    [FEATURE] Properly commented all methods in code

commit 7ee45df92cffc782f0e49984dca1ccfeb7dc93b9
Author: Zach Davis <zach@castironcoding.com>
Date:   Fri Oct 21 11:49:14 2011 -0700

    [FEATURE] Removed the ability to add layouts. FluidPage templates are whole HTML documents, so it doesn't make sense to wrap additional layout around the output

commit 5d7e74e00f3eab7fb7e74526dac1edb2984c3b38
Author: Zach Davis <zach@castironcoding.com>
Date:   Fri Oct 21 11:35:57 2011 -0700

    [FEATURE] Refactored code to be less procedural and more in keeping with current TYPO3 dev practices
    
    Moved code out of pi1 and generally cleaned up the codebase. After looking at the fluidtemplate implemenation again, I tweaked some typoscript to keep things more in sync with fluidtemplate.

commit 1b8d3d0c4c298aca6d61082be3c064866feaa26c
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 14:10:57 2011 -0700

    [CHORE] Commiting after TER upload to reflect change to ext_emconf.php

commit a1dbf2bb11f25e7f82a2fa9a1657402e5f4e5500
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 14:08:44 2011 -0700

    [CHORE] Added extension icon and updated permissions on autoload file

commit 797df262a8035ffc0b1b2b3e5e5504c0e307ba68
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 13:49:48 2011 -0700

    [FEATURE] Added layout constants and global constants to the rendering of the fluid template

commit aaa6b627483ce6f844fadf15bf9c77ed6600802b
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 13:49:12 2011 -0700

    [BUGFIX] Removed body class field that FluidPage was adding to backend layout reocrds

commit b3c4d1a51eca13e9a487260361c96b6fd8b579fa
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 13:48:34 2011 -0700

    [CHORE] Removed old sample template assets that were no longer needed

commit 94ddb31f1e294c357df33d9099a141edaee2d621
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 13:48:07 2011 -0700

    [FEATURE] Added rootline field config to ext_localconf.php

commit 98231a260c228e50e6aca23fd036a245a4d17343
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 13:47:06 2011 -0700

    [FEATURE] Wrote basic documentation

commit b11fb150c4669f6e6eb1b48292889d3f3d0fc3c7
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 13:46:41 2011 -0700

    [CHORE] Removed the wizard files in doc

commit c9f3eee522b5d022369f0fbc7d0c41b880c2cba2
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 13:46:11 2011 -0700

    [FEATURE] Added a sample HTML template

commit 94c9b88e7b3dd151ccd4e242862b0a77d39d8694
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 11:05:29 2011 -0700

    [FEATURE] Removed the body class field from backend layouts; this can be handled with typoscript

commit 236cf757455d62d46d69b4ecb3cc4a863ef9cc11
Author: Zach Davis <zach@castironcoding.com>
Date:   Thu Oct 20 09:41:59 2011 -0700


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "fluidpage".
#
# Auto generated 15-11-2011 10:26
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Fluid Page Templates',
	'description' => 'This extension leverages the TYPO3 backend layout record and Fluid typoscript templates to make editing more intuitive',
	'category' => 'fe',
	'author' => 'Cast Iron Coding, Inc',
	'author_email' => 'zach@castironcoding.com',
	'shy' => '',
	'dependencies' => 'fluid',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'stable',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.2.0',
	'constraints' => array(
		'depends' => array(
			'fluid' => '',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:39:{s:9:"ChangeLog";s:4:"7c38";s:10:"README.txt";s:4:"12b2";s:16:"ext_autoload.php";s:4:"109f";s:12:"ext_icon.gif";s:4:"8343";s:17:"ext_localconf.php";s:4:"955f";s:14:"ext_tables.php";s:4:"fb7a";s:31:"Classes/Controller/Template.php";s:4:"f34a";s:26:"Classes/Model/Template.php";s:4:"0ef2";s:35:"Classes/Service/TemplateFactory.php";s:4:"1f52";s:41:"Classes/ViewHelpers/ContentViewHelper.php";s:4:"5a5f";s:30:"Resources/Public/css/style.css";s:4:"a763";s:31:"Resources/Public/html/main.html";s:4:"3909";s:30:"Resources/Public/images/bg.gif";s:4:"20be";s:42:"Resources/Public/images/bg_contentWrap.gif";s:4:"9edc";s:41:"Resources/Public/images/bg_firstLevel.gif";s:4:"1186";s:37:"Resources/Public/images/bg_footer.jpg";s:4:"1826";s:41:"Resources/Public/images/bg_getStarted.jpg";s:4:"8415";s:37:"Resources/Public/images/bg_header.jpg";s:4:"b45b";s:40:"Resources/Public/images/bg_liCurrent.gif";s:4:"d6cd";s:43:"Resources/Public/images/bg_searchButton.gif";s:4:"a13d";s:42:"Resources/Public/images/bg_searchInput.gif";s:4:"0fdf";s:45:"Resources/Public/images/bg_secondaryLevel.jpg";s:4:"62ad";s:47:"Resources/Public/images/button_logintotypo3.png";s:4:"69bf";s:40:"Resources/Public/images/flower_mask2.png";s:4:"2c56";s:42:"Resources/Public/images/header_overlay.png";s:4:"501b";s:44:"Resources/Public/images/navButtonOverlay.png";s:4:"4d85";s:38:"Resources/Public/images/navOverlay.png";s:4:"73f1";s:33:"Resources/Public/images/step1.png";s:4:"748b";s:33:"Resources/Public/images/step2.png";s:4:"7317";s:33:"Resources/Public/images/step3.png";s:4:"4070";s:13:"doc/howto.rst";s:4:"edf0";s:14:"doc/styles.css";s:4:"1adc";s:20:"doc/build/index.html";s:4:"280c";s:30:"pi1/class.tx_fluidpage_pi1.php";s:4:"852e";s:53:"static/fluid_page_sample_page_templates/constants.txt";s:4:"f6fe";s:49:"static/fluid_page_sample_page_templates/setup.txt";s:4:"dabd";s:58:"static/fluid_page_static_typoscript_template/constants.txt";s:4:"9363";s:54:"static/fluid_page_static_typoscript_template/setup.txt";s:4:"660f";s:32:"static/tsconfig/pageTSConfig.txt";s:4:"84af";}',
	'suggests' => array(
	),
);

?>